Sming/Libraries/ModbusMaster/samples/generic/app/application.cpp

```diff
@@ -0,0 +1,126 @@
+#include <SmingCore.h>
+#include <Debug.h>
+#include <ModbusMaster.h>
+
+#define MODBUS_COM_SPEED 115200
+#define RS485_RE_PIN 15
+#define MB_SLAVE_ADDR 1
+#define SLAVE_REG_ADDR 1
+
+Timer mbLoopTimer;
+
+ModbusMaster mbMaster;
+HardwareSerial modbusComPort(UART0);
+HardwareSerial debugComPort(UART1);
+
+uint16_t globalSeconds = 0;
+
+void mbLoop()
+{
+	globalSeconds++;
+
+	uint8_t numberOfRegistersToWrite = 1;
+	uint8_t bufferPosition = 0;
+	mbMaster.begin(MB_SLAVE_ADDR, Serial);
+	mbMaster.setTransmitBuffer(bufferPosition, globalSeconds);
+	mbMaster.writeMultipleRegisters(SLAVE_REG_ADDR, numberOfRegistersToWrite);
+
+	uint8_t nrOfRegistersToRead = 1;
+	uint8_t mbResult = mbMaster.readHoldingRegisters(SLAVE_REG_ADDR, nrOfRegistersToRead); //see also readInputRegisters
+	uint16_t result = 0;
+
+	if(mbResult == mbMaster.ku8MBSuccess) {
+		/*
+		for(i = 0; i < nrOfRegisters; i++) {
+			buffer[i] = mbMaster.getResponseBuffer(i);
+			debugComPort.printf("Reg %d: %d\r\n", i, buffer[i]);
+		}
+		*/
+		debugf("Data from slave: %d", mbMaster.getResponseBuffer(0));
+	} else {
+		debugf("Res err: %d", result);
+	}
+
+	mbMaster.clearResponseBuffer();
+}
+
+void preTransmission()
+{
+	digitalWrite(RS485_RE_PIN, HIGH);
+	delayMilliseconds(2);
+}
+
+void postTransmission()
+{
+	delayMicroseconds(500);
+	digitalWrite(RS485_RE_PIN, LOW);
+}
+
+void mbLogReceive(const uint8_t* adu, size_t aduSize, uint8_t status)
+{
+	if(status != mbMaster.ku8MBSuccess) {
+		switch(status) {
+		case mbMaster.ku8MBIllegalFunction:
+			debugf("MB Illegal Function");
+			break;
+		case mbMaster.ku8MBIllegalDataAddress:
+			debugf("MB Illegal Address");
+			break;
+		case mbMaster.ku8MBIllegalDataValue:
+			debugf("MB Illegal Data Value");
+			break;
+		case mbMaster.ku8MBSlaveDeviceFailure:
+			debugf("MB Slave Device Failure");
+			break;
+		case mbMaster.ku8MBInvalidSlaveID:
+			debugf("MB Invalid Slave ID");
+			break;
+		case mbMaster.ku8MBInvalidFunction:
+			debugf("MB Invalid function");
+			break;
+		case mbMaster.ku8MBResponseTimedOut:
+			debugf("MB Response Timeout");
+			break;
+		case mbMaster.ku8MBInvalidCRC:
+			debugf("MB Invalid CRC");
+			break;
+		case mbMaster.ku8MBResponseTooLarge:
+			debugf("MB Response too large");
+			break;
+		}
+		debugf("ADU Size: %d, status: %d ", aduSize, status);
+	}
+	debugf("\r\n");
+}
+
+void mbLogTransmit(const uint8_t* adu, size_t aduSize)
+{
+	uint8_t i = 0;
+	while(i < aduSize) {
+		debug_hex(INFO, "ADU", adu, aduSize);
+		i++;
+	}
+	debugf(" ");
```

Comment:
LOL. You can replace the entire function with `debug_hex(INFO, "ADU", adu, aduSize);`. Produces output like this (excerpt from HostTests which you can see in the CI Test logs):

```
encode output: 41 47 52 76 62 6d 74 6c 65 51 42 72 61 57 35 6e  AGRvbmtleQBraW5n
               63 47 6c 75 00                                   cGlu.
```

If this isn't quite what you want there are some configurable parameters, just a suggestion though. The `debug_hex` macro is there when you only want it emitted at a certain debug level, or you can use `m_printhex()` directly.

Reply:
Ah I see... Btw hold a moment, the app doesn't work on the hardware.. Investigating.

Reply:
Wow, debug_hex did a great job! Just committed last fixes. What concerns me is that the ```ku16MBResponseTimeout``` is set to 2000 ms in ModbusMaster.h and it triggers WDT reset. Should I provide a patch with more suitable timeout for Sming?

Reply:
If the sample doesn't work then yes, needs a fix.

Reply:
The sample works when a slave device is connected. If it isn't however, WDT reset...

Reply:
Hanging around in a look for that length of time is a bad idea. For the purposes of this sample, can you make the timeout configurable?